Adding debugger awarness with Keil MDK

deepikabhavnani opened the pull request:
### Description
Main thread in Mbed OS is statically allocated and was not available in call stack of Keil MDK. The RTX5 kernel requires statically allocated thread information objects that are placed into a specific section to enable RTOS thread awareness in Keil MDK. This fix is to keep main thread in specific section of memory.

Issue: https://github.com/ARMmbed/mbed-os/issues/8793

### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Commit:
Adding debugger awarness with Keil MDK

Main thread in Mbed OS is statically allocated and was not available in call
stack of Keil MDK. The RTX5 kernel requires statically allocated thread
information objects that are placed into a specific section to enable RTOS
thread awareness in Keil MDK. This fix is to keep main thread in specific
section of memory.

kjbracey commented:
How does this interact with #7244? If that still applies, this will cost a lot of ROM in ARMCC?

Presumably nothing you could do about the underlying issue, but maybe you could restrict the section attribute to `MBED_DEBUG` builds?

deepikabhavnani commented:
Tested blinky example with and without change to see the impact on ROM size, please see the results below:

Without Fix
```
| Module                                                       |         .text |       .data |        .bss |
|--------------------------------------------------------------|---------------|-------------|-------------|
| mbed-os\rtos\TARGET_CORTEX\mbed_rtos_rtx.o                   |     216(+216) |       4(+4) | 4248(+4248) |

| Subtotals                                                    | 39461(+39461) | 2549(+2549) | 8292(+8292) |
Total Static RAM memory (data + bss): 10841(+10841) bytes
Total Flash memory (text + data): 42010(+42010) bytes

```
With Fix
```
| Module                                                       |         .text |       .data |        .bss |
| mbed-os\rtos\TARGET_CORTEX\mbed_rtos_rtx.o                   |   216(+0)     |   72(+68)   | 4180(-68)   |
| Subtotals                                                    | 39517(+56)    | 2617(+68)   | 8224(-68)   |
Total Static RAM memory (data + bss): 10841(+0) bytes
Total Flash memory (text + data): 42134(+124) bytes
```

deepikabhavnani commented:
@kjbracey-arm - Having just the control block in BSS section resolves the issue, i.e. we are not putting the main thread stack in .bss here. https://github.com/ARMmbed/mbed-os/pull/8887/files#diff-91e672466b09fb21b174f0f44e07b8aaL31 hence the size impact is not much.

On another thought can we do same do Idle / timer threads?
```

C:\Users\deebha02\work_mbed\KeilIDE_fix\mbed-os>git diff
diff --git a/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c b/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c
index 8fe704f80..f35481981 100644
--- a/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c
+++ b/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c
@@ -122,8 +122,7 @@ static osRtxThread_t os_idle_thread_cb \
 __attribute__((section(".bss.os.thread.cb")));

 // Idle Thread Stack
-static uint64_t os_idle_thread_stack[OS_IDLE_THREAD_STACK_SIZE/8] \
-__attribute__((section(".bss.os.thread.stack")));
+static uint64_t os_idle_thread_stack[OS_IDLE_THREAD_STACK_SIZE/8];

 // Idle Thread Attributes
 static const osThreadAttr_t os_idle_thread_attr = {
@@ -179,8 +178,7 @@ static osRtxThread_t os_timer_thread_cb \
 __attribute__((section(".bss.os.thread.cb")));

 // Timer Thread Stack
-static uint64_t os_timer_thread_stack[OS_TIMER_THREAD_STACK_SIZE/8] \
-__attribute__((section(".bss.os.thread.stack")));
+static uint64_t os_timer_thread_stack[OS_TIMER_THREAD_STACK_SIZE/8];

```
`
| mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\rtx_lib.o                                                               |   285(+0) |  332(-768) | 1280(+768) |
`

kjbracey commented:
We could make such a local change to RTX, #7244 closed with a change not happening upstream, and not making a local change either, but we could still do something locally.

And just removing the area directive from the stacks altogether seems plausible - if we have no special handling for the .bss.os area in our linker maps.

If this PR was a net ROM reduction by taking the stacks out of ROM and putting this control block in, @bulislaw  couldn't complain.

deepikabhavnani commented:
> if we have no special handling for the .bss.os area in our linker maps.

I am not sure of any special handling for .bss.os. 
@c1728p9 - your inputs please..

Commit:
CMSIS/RTX: Move Idle and Timer thread stack to bss section.

In case of ARM compiler, idle and timer thread stack though assigned
to `.bss.os` section since not zero initialized are part of `data` section.

In this commit, we are moving stacks of idle and timer thread to bss
section and thereby saving ROM space.

deepikabhavnani commented:
@JonatanAntoni - Please review and share your thoughts on this PR

JonatanAntoni commented:
Hi @deepikabhavnani,

this looks meaningful to me. The relevant part for RTX5 awareness in MDK (and potentially other IDEs) is to place all the object control blocks to their specific memory sections. This applies to all RTOS objects (not only threads) when control blocks are allocated manually (from RTX point of view).

Cheers,
Jonatan

deepikabhavnani commented:
@c1728p9 @bulislaw - Please review

c1728p9 commented on tools/profiles/debug.json:
```diff
@@ -23,7 +23,7 @@
         "c": ["-D__ASSERT_MSG", "-std=gnu99"],
         "cxx": ["-fno-rtti", "-std=gnu++98"],
-        "ld": ["--verbose", "--remove", "--show_full_path", "--legacyalign"]
+        "ld": ["--verbose", "--remove", "--show_full_path", "--legacyalign", "--keep=os_cb_sections"]
```
@jreineckearm should this be `.bss.os.thread.cb` rather than `os_cb_sections`?

Also, the control block, ```osRtxThread_t```, is passed into ```osThreadNew``` as part of ```osThreadAttr_t``` so it should be referenced. Is the ```--keep``` actually needed with more recent versions of RTX?

c1728p9 commented:
> The relevant part for RTX5 awareness in MDK (and potentially other IDEs) is to place all the object control blocks to their specific memory sections. This applies to all RTOS objects (not only threads) when control blocks are allocated manually (from RTX point of view).

@JonatanAntoni does this mean that threads created by using the `Thread` class in mbed-os will not show up in MDKs RTX5 awareness? The Thread class contains the control block as [one of its members](https://github.com/ARMmbed/mbed-os/blob/master/rtos/Thread.h#L540) - `_obj_mem`. Note that `mbed_rtos_storage_thread_t` is a typedef of `osRtxThread_t`.

JonatanAntoni commented:
> [..] does [it] mean that threads created by using the `Thread` class in mbed-os will not show up in MDKs RTX5 awareness?

I have never tried to debug a mbed application using MDK, but I assume it won't show up. The RTOS-awareness (generally speaking component awareness) in MDK is generated using a generic [Software Component View Description](https://www.keil.com/pack/doc/compiler/EventRecorder/html/SCVD_Format.html). The section names are used as a common entry point for control block look up. Control blocks spread randomly over the memory cannot be found. 

> Is the --keep actually needed with more recent versions of RTX?

--keep is not needed for Arm Compiler 6 but for Arm Compiler 5. The symbol `os_cb_sections` denotes the top level entry for control block look up, see [rtx_lib.c:583](https://github.com/ARM-software/CMSIS_5/blob/2efcd747d7cc1c09ab9f5f430139635f611ebcbe/CMSIS/RTOS2/RTX/Source/rtx_lib.c#L583).

Cheers,
Jonatan

cmonr commented:
Testing PR with CI.

deepikabhavnani commented:
@c1728p9 @kjbracey-arm @bulislaw - Please review

bulislaw commented on rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c:
```diff
@@ -122,9 +122,12 @@ static osRtxThread_t os_idle_thread_cb \
 __attribute__((section(".bss.os.thread.cb")));
 
 // Idle Thread Stack
+#if defined (__ARMCC_VERSION)
```
Two things:
* I'm assuming that checks whether it's 5 or 6, but it's not obvious which way, could you add comment? Also why are we only doing it for one compiler should be documented.
* Is this commit accepted into CMSIS repo? If so please reference it in the commit message. I really want to avoid modifying RTX/CMSIS in Mbed.

deepikabhavnani replied:
Good catch! it should be compiler check and not version of compiler. Will correct that,

> Is this commit accepted into CMSIS repo? I

I don't think it will be acceptable into CMSIS repo, but it saves us some memory in flash.

Commit:
Add SHA for "Move Idle and Timer thread stack to bss section" commit

cmonr commented:
CI started

deepikabhavnani commented:
Conflicts :-( 
@ARMmbed/mbed-os-maintainers - This PR is tagged for 5.12 and was ready to merge, do we have any restrictions on merge to master for PR's tagged for minor release?

deepikabhavnani commented:
Rebased and conflicts resolved.  Changing label from `ready to merge` to `needs:CI`

cmonr commented:
@deepikabhavnani Nothing hard, but we definitely prefer if PRs for minior releases come in a bit later than right after a minor release is made, if that makes sense. 

Lowers the risk that early 5.x.{1,2,3} patch releases will be troublesome.

cmonr commented:
CI started

0xc0170 commented:
Client test got stuck, restarted

deepikabhavnani commented:
> we definitely prefer if PRs for minior releases come in a bit later than right after a minor release is made, if that makes sense.

👍  This helps, will consider this when adding changes for minor release. Else can we add and park those changes ?
